### PR TITLE
Improve `db reingest range` command

### PR DESCRIPTION
**Pull request by bartekn**
Improves `db reingest range` command by starting multiple workers at the same time and committing work more often. Introduces `util.WorkersPool ` responsible for distributing work between workers.

This changes some queries that required Postgres 9.5+ (upsert query) so a base branch was set to `release-horizon-v0.18.0`.

Fix (partially): #931 and #917.

### CODE REVIEW

**Comment by howardtw** [comment]
Do you mind rebasing this PR and resolve the conflicts?

**Comment by bartekn** [comment]
@howardtw rebased. Can you take a look again?

**Comment by howardtw** [comment]
I wonder if we can break this PR into two different ones: one for introducing workers, one for making `from` & `to` CLI flags, since most of the work here is aiming to add worker functionalities.

**Comment by bartekn** [comment]
> I wonder if we can break this PR into two different ones

Yeah, maybe that would make sense before a review but now, after 2 rounds, I think it doesn't make sense.

**Comment by howardtw** [comment]
Well the first round was to ask for resolving merge conflict. That was really a review.

**Comment by bartekn** [comment]
OK, you're right. I was quickly switching between PR and haven't noticed it. I think I'm good with splitting this into 2 PRs.

**Comment by bartekn** [comment]
@howardtw done, CLI params will be added in a separate PR (not sure if this is the final form of the code - pre review).

File: services/horizon/internal/util/workers_pool.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	work   []interface{}
+	worker func(workerID int, work interface{})
+	mutex  sync.Mutex
+}
+
+func (w *WorkersPool) AddWork(job interface{}) {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	if w.work == nil {
```
[the remark this answers is not in the document]
Comment (bartekn) [reply services/horizon/internal/util/workers_pool.go]: Thanks, didn't know about it.

File: services/horizon/internal/util/workers_pool.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+func (w *WorkersPool) WorkSize() int {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	return len(w.work)
+}
+
+// getWork returns a work. The second return value is false if there's no more work.
+func (w *WorkersPool) getWork() (interface{}, bool) {
```
[the remark this answers is not in the document]
Comment (bartekn) [reply services/horizon/internal/util/workers_pool.go]: I don't remember reasoning behind it. Maybe I was thinking that `slice[1:]` rewrites a slice (it doesn't). Changed it to getting a front element.

File: services/horizon/cmd/db.go [hunk 3]
```diff
@@ -365,3 +367,66 @@ func reingest(cmd reingestType, args ...int32) {
...
+			"to":   lr.to,
+		})
+
+		localLog.Info("Worker starting range...")
+
+		_, err := i.ReingestRange(lr.from, lr.to)
+		if err != nil {
+			localLog.WithField("err", err).Error("Worker failed range, work will processed again")
+			// Add the work again
+			pool.AddWork(lr)
```
[the remark this answers is not in the document]
Comment (bartekn) [reply services/horizon/cmd/db.go]: No, all queries to a DB are sent in a transaction.

File: services/horizon/cmd/db.go [hunk 4]
```diff
@@ -365,3 +367,66 @@ func reingest(cmd reingestType, args ...int32) {
...
+			os.Exit(1)
+		}
+
+		localLog := hlog.WithFields(hlog.F{
+			"id":   workerID,
+			"from": lr.from,
+			"to":   lr.to,
+		})
+
+		localLog.Info("Worker starting range...")
```
[the remark this answers is not in the document]
Comment (bartekn) [reply services/horizon/cmd/db.go]: These values are appended to `localLog` above. No need to duplicate it.

File: services/horizon/cmd/db.go [hunk 5]
```diff
@@ -365,3 +367,66 @@ func reingest(cmd reingestType, args ...int32) {
...
+
+		_, err := i.ReingestRange(lr.from, lr.to)
+		if err != nil {
+			localLog.WithField("err", err).Error("Worker failed range, work will processed again")
```
[the remark this answers is not in the document]
Comment (howardtw) [reply services/horizon/cmd/db.go]: There is a typo here -> `work will be processed again`

**Comment by bartekn** [comment]
@howardtw fixed. PR branch was way behind Circle CI integration that's why it didn't work.

**Pull request merged**